src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/EnumerableConditionalWeakTable.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Roslyn.Utilities;
+
+#if NET
+internal readonly struct EnumerableConditionalWeakTable<TKey, TValue>() : IEnumerable<KeyValuePair<TKey, TValue>>
```

Comment:
*alterantively*.  instead of having this be a struct, which makes things like locking on it confusing later on (needing another object), it seems like this can just be an alias.  

Oh.  crap.  it's generic.  ANNOYING.  ok.  i see why you did this.

Reply:
consider having this expose an `object SyncLock => ...` property that returns the underlyign CWT.  the consumign code can do `lock (cwt.SyncLock)`.  it would prevent an allocation later on.  but i don't know if you care.

Reply:
Yeah... Looking forward extension types :)